From c6b0884c343f42001b8d2a564b8edba6000a714a Mon Sep 17 00:00:00 2001
From: Jakob Spahn <jakob@craalse.de>
Date: Wed, 11 Dec 2024 16:17:58 +0100
Subject: [PATCH 1/5] makefile generator

---
 ocelot/CMakeLists.txt | 48 ++++++++++++++++++++++++++++++++++---------
 1 file changed, 38 insertions(+), 10 deletions(-)

diff --git a/ocelot/CMakeLists.txt b/ocelot/CMakeLists.txt
index 30981b52..9cd4d012 100644
--- a/ocelot/CMakeLists.txt
+++ b/ocelot/CMakeLists.txt
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
 
-project(gpuocelot C CXX CUDA ASM)
+project(gpuocelot C CXX ASM)
 
 option(BUILD_LLVM "Use LLVM compiled from sources: default OFF" OFF)
 option(BUILD_TESTS "Build tests: default OFF" OFF)
@@ -20,14 +20,25 @@ list(APPEND CUDA_NVCC_FLAGS "-Wno-deprecated-gpu-targets -Wno-deprecated-declara
 
 set(CMAKE_CUDA_SEPARABLE_COMPILATION ON)
 
+
+set(BISON_EXECUTABLE /opt/homebrew/opt/bison/bin/bison)  # native version is outdated
+
 find_package(Boost COMPONENTS system filesystem REQUIRED)
 find_package(FLEX 2.5 REQUIRED)
 find_package(BISON 2.5 REQUIRED)
 find_package(GLEW REQUIRED)
+find_library(GLEW_LIBRARY libGLEW.dylib)
 find_package(ZLIB REQUIRED)
 find_library(ZSTD_LIB NAMES zstd libzstd)
 
+find_library(LIBEDIT_LIB edit REQUIRED)
+find_library(LIBFFI_LIB ffi REQUIRED)
+find_library(LIBNCURSES_LIB ncurses REQUIRED)
+find_library(LIBXAR_LIB xar REQUIRED)
+find_library(LIBXML2_LIB xml2 REQUIRED)
+
 if ("x${BUILD_LLVM}" STREQUAL "xOFF")
+set(LLVM_DIR "/opt/homebrew/opt/llvm@15/lib/cmake/llvm")
 find_package(LLVM REQUIRED CONFIG)
 
 file(GLOB LLVM_LIBRARIES ${LLVM_LIBRARY_DIR}/*.a)
@@ -230,12 +241,11 @@ set(${PROJECT_NAME}_INCLUDE_DIRS
 	${CMAKE_CURRENT_BINARY_DIR})
 
 set(${PROJECT_NAME}_LINK_LIBRARIES
-	${GLEW_LIBRARIES}
+	${GLEW_LIBRARY}
 	hydrazine
 	Boost::system
 	Boost::filesystem
 	ZLIB::ZLIB
-	tinfo
 	${ZSTD_LIB}
 	${CMAKE_DL_LIBS})
 
@@ -528,24 +538,42 @@ target_link_directories(${PROJECT_NAME}_util PUBLIC ${${PROJECT_NAME}_LIBRARY_DI
 add_library(${PROJECT_NAME} SHARED src/ocelot.cpp)
 target_include_directories(${PROJECT_NAME} PUBLIC ${${PROJECT_NAME}_INCLUDE_DIRS})
 target_link_libraries(${PROJECT_NAME} PRIVATE
-	"-Wl,--whole-archive"
+	"-Wl,-force_load"
 	${PROJECT_NAME}_analysis
+	"-Wl,-force_load"
 	${PROJECT_NAME}_api
+	"-Wl,-force_load"
 	${PROJECT_NAME}_cal
+	"-Wl,-force_load"
 	${PROJECT_NAME}_cuda
+	"-Wl,-force_load"
 	${PROJECT_NAME}_executive
+	"-Wl,-force_load"
 	${PROJECT_NAME}_ir
+	"-Wl,-force_load"
 	${PROJECT_NAME}_parser
+	"-Wl,-force_load"
 	${PROJECT_NAME}_trace
+	"-Wl,-force_load"
 	${PROJECT_NAME}_transforms
+	"-Wl,-force_load"
 	${PROJECT_NAME}_translator
-	${PROJECT_NAME}_util
-	"-Wl,--no-whole-archive")
+	"-Wl,-force_load"
+	${PROJECT_NAME}_util)
+
+target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBEDIT_LIB})
+target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBFFI_LIB})
+target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBNCURSES_LIB})
+target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBXAR_LIB})
+target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBXML2_LIB})
+target_link_libraries(${PROJECT_NAME} PUBLIC "-framework CoreFoundation")
+target_link_libraries(${PROJECT_NAME} PUBLIC "-framework CoreServices")
+
 target_link_libraries(${PROJECT_NAME} PUBLIC ${${PROJECT_NAME}_LINK_LIBRARIES})
-target_link_libraries(${PROJECT_NAME} PRIVATE -Wl,--gc-sections -Wl,--start-group ${LLVM_LIBRARIES} -Wl,--end-group)
+target_link_libraries(${PROJECT_NAME} PRIVATE -Wl,-dead_strip ${LLVM_LIBRARIES})
 target_link_directories(${PROJECT_NAME} PUBLIC ${${PROJECT_NAME}_LIBRARY_DIRS})
-target_link_libraries(${PROJECT_NAME} PUBLIC ${ZLIB_LIBRARIES} tinfo)
-install(TARGETS ${PROJECT_NAME} LIBRARY DESTINATION /usr/lib)
+target_link_libraries(${PROJECT_NAME} PUBLIC ${ZLIB_LIBRARIES})
+install(TARGETS ${PROJECT_NAME} LIBRARY DESTINATION /usr/local/lib)
 install(CODE "execute_process(COMMAND ldconfig)")
 
 if(BUILD_TOOLS)
@@ -572,7 +600,7 @@ endif()
 
 if(BUILD_TESTS)
 	ocelot_add_tests(analysis)
-	ocelot_add_tests(api)
+	#ocelot_add_tests(api)
 	ocelot_add_tests(cal)
 	ocelot_add_tests(executive)
 	ocelot_add_tests(ir)
-- 
2.39.5 (Apple Git-154)

