Submodule ocelot/ThirdParty/res_embed contains modified content
diff --git a/ocelot/ThirdParty/res_embed/cmake/ResEmbed.cmake b/ocelot/ThirdParty/res_embed/cmake/ResEmbed.cmake
index 3224975..0a46012 100644
--- a/ocelot/ThirdParty/res_embed/cmake/ResEmbed.cmake
+++ b/ocelot/ThirdParty/res_embed/cmake/ResEmbed.cmake
@@ -12,7 +12,7 @@ function(res_embed)
 	set(RES_EMBED_CURRENT_INCLUDE_DIR ${_RES_EMBED_PATH}/../include)
 
 	if(NOT DEFINED USE_NASM)
-		if (MSVC OR APPLE)
+		if (MSVC)
 			set(USE_NASM TRUE)
 		else()
 			set(USE_NASM FALSE)
diff --git a/ocelot/ThirdParty/res_embed/include/res_embed.gas.in b/ocelot/ThirdParty/res_embed/include/res_embed.gas.in
index 3c66b05..cbccda2 100644
--- a/ocelot/ThirdParty/res_embed/include/res_embed.gas.in
+++ b/ocelot/ThirdParty/res_embed/include/res_embed.gas.in
@@ -1,6 +1,6 @@
 # Embed content into hex array for in-memory caching
 
-.section .data
+.data
 .global @OS_DEPENDENT_PREFIX@hex_content_@FILE_KEY_HASH@
 
 @OS_DEPENDENT_PREFIX@hex_content_@FILE_KEY_HASH@:
@@ -9,4 +9,4 @@
 .global @OS_DEPENDENT_PREFIX@hex_size_@FILE_KEY_HASH@
 
 @OS_DEPENDENT_PREFIX@hex_size_@FILE_KEY_HASH@:
-.int @OS_DEPENDENT_PREFIX@hex_size_@FILE_KEY_HASH@ - @OS_DEPENDENT_PREFIX@hex_content_@FILE_KEY_HASH@
+.quad (@OS_DEPENDENT_PREFIX@hex_size_@FILE_KEY_HASH@ - @OS_DEPENDENT_PREFIX@hex_content_@FILE_KEY_HASH@)
